File: GFF/Covering time/CT.c
#include<stdio.h>
#include<stdlib.h>
#include<time.h>
#include<math.h>
#define PI acos(-1)
//#define _DEBUG
void Torus_Cov(int* T_cov, int n, int iter, int debug){
// Ce script calcule une sequence de temp couverture en une torus annulus T_N par MCMC

    srand(7);
    int L_t[n][n];                      // Local time matrix
    for(int i = 0;i < iter; i++){
        if(i % 10000 == 0)
            printf("%d iteration processed!\n", i);
        int X_t     = 0;
        int Y_t     = 0;
        for(int j = 0; j < n; j++)
            for(int k = 0; k < n; k++)
                L_t[j][k] = 0;
        int n_V     = n * n;                // # of not-v node
        int t       = 0;
        while(n_V != 0){
            if(L_t[X_t][Y_t] == 0)
                n_V = n_V - 1;
            L_t[X_t][Y_t]   += 1;
            t   = t + 1;
            int rnd = rand() % 4;
            if(rnd == 3){
                #ifdef _DEBUG
                    printf("right\n");
                #endif
                X_t = (X_t + 1) % n;
            }
            else if(rnd == 2){
                #ifdef _DEBUG
                    printf("left\n");
                #endif
                X_t = (X_t - 1 + n) % n;
            }
            else if(rnd == 1){
                #ifdef _DEBUG
                    printf("up\n");
                #endif
                Y_t = (Y_t + 1) % n;
            }
            else{
                #ifdef _DEBUG
                    printf("down\n");
                #endif
                Y_t = (Y_t - 1 + n) % n;
            }
        }
        T_cov[i]    = t;
    }
}

int main(){
    time_t start,end;
    start = time(NULL);
    int n = 10, iter = 10000000;
    int* T_cov;
    T_cov = (int *)malloc(sizeof(int) * (iter + 1));
    Torus_Cov(T_cov, n, iter, 0);
    end = time(NULL);		 
    printf("time = %lf\n",difftime(end,start));

    // write the data
    int write = 0;
    if(write == 1){
        FILE* fp = fopen("C50.txt","ab+");
        for(int i = 0; i < iter; i++)
            fprintf(fp, "%d\n", T_cov[i]);
        fclose(fp);
    }
    free(T_cov);
    return 0;
}

File: GFF/Covering time/CT.py
import numpy as np
import numpy.linalg as alg
import numpy.random as r
import scipy.linalg as alg
import scipy.sparse as spa
from time import time

def Torus_Cov( n, iter ):
# Ce script calcule une sequence de temp couverture en une torus annulus 
# T_N par MCMC

    # Parameters for random walk
    X_t     = 0
    Y_t     = 0
    T_cov   = np.zeros(iter)

    for i in range(iter):
        L_t     = np.zeros([n, n])            # Local time matrix
        n_V     = n * n                       # of not-v node
        t       = 0
        while n_V != 0:
            if(L_t[X_t, Y_t] == 0):
                n_V = n_V - 1
                L_t[X_t, Y_t] = 1
            t   = t + 1
            rnd = r.rand(1)
            if rnd > 0.75:
                X_t = np.mod(X_t + 1, n)
            elif rnd > 0.5:
                X_t = np.mod(X_t - 1, n)
            elif rnd > 0.25:
                Y_t = np.mod(Y_t + 1, n)
            else:
                Y_t = np.mod(Y_t - 1, n)
        print(t / np.power(n * np.log(n), 2) * np.pi / 4)
        T_cov[i]    = t
        
    return T_cov


start = time()
n = 50
iter = 100
T_cov = Torus_Cov(n, iter)
print(time() - start)
write = False
if write:
    with open('50.txt','a',encoding='utf-8') as f:#使用with open()新建对象f
        for i in range(iter):
            f.write(str(T_cov[i]))
            f.write('\n')